index.bs

```diff
@@ -4109,165 +4310,1238 @@ To <dfn>get the response data</dfn> given |response|:
+         URL Pattern=] with |pattern|["<code>port</code>"].
+
+      1. Append "<code>:</code>" to |pattern url|.
+
+      1. For each |codepoint| in |port|:
+
+        1. If |codepoint| is not an [=ASCII digit=]:
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |port| to |pattern url|.
+
+    1. If |pattern| [=map/contains=] "<code>pathname</code>":
+
+      1. Let |pathname| be the result of [=trying=] to [=unescape URL Pattern=] with
+         |pattern|["<code>pathname</code>"].
+
+      1. If |pathname| does not [=string/starts with|start with=] U+002F (/), then
+         append "<code>/</code>" to pattern.
+
+      1. For each |codepoint| in |pathname|:
+
+        1. If «U+003F (?), U+0023 (#)» [=set/contains=]
+           |codepoint|:
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |pathname| to |pattern url|.
+
+    1. If |pattern| [=map/contains=] "<code>search</code>":
+
+      1. Let |search| be the result of [=trying=] to [=unescape URL pattern=] with
+         |pattern|["<code>search</code>"].
+
+      1. If |search| does not [=string/starts with|start with=] U+003F (?), then
+         append "<code>?</code>" to |pattern url|.
+
+      1. For each |codepoint| in |search|:
+
+        1. If |codepoint| is U+0023 (#):
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |search| to |pattern url|.
+
+1. Otherwise, if |pattern| matches the <code>network.UrlPatternString</code> production:
+
+  1. Let |pattern url| be the result of [=trying=] to [=unescape URL
+     pattern=] with |pattern|["<code>pattern</code>"].
+
+1. Let |url| be the result of [=URL Parser|parsing=] |pattern url|.
+
+1. If |url| is failure, return [=error=] with [=error code=] [=invalid argument=].
+
+1. Let |parsed| be a struct with the following fields:
+   <dl>
+     <dt>protocol
+     <dd>|url|'s [=url/scheme=] if |has protocol| is true, or null otherwise.
+     <dt>hostname
+     <dd>|url|'s [=url/host=] if |has hostname| is true, or null otherwise.
+     <dt>port
+     <dd>[=Serialize an integer=] with |url|'s [=url/port=], if it's not null,
+         or null otherwise.
+     <dt>pathname
+     <dd>The result of running the [=URL path serializer=] with |url|, if
+         |url|'s [=path=] is not the empty string and is not [=list/empty=],
+         or null otherwise.
+     <dt>search
+     <dd>|url|'s [=url/query=].
+   </dl>
+
+1. Return [=success=] with data |parsed|.
+
+</div>
+
+<div algorithm>
+To <dfn>match URL pattern</dfn> given |url pattern| and |url string|:
+
+1. Let |url| be the result of [=URL parser|parsing=] |url string|.
+
+1. If |url pattern|'s protocol is not null and is not equal to |url|'s
+   [=url/scheme=], return false.
+
+1. If |url pattern|'s hostname is not null and is not equal to |url|'s
+   [=url/host=], return false.
+
+1. If |url pattern|'s port is not null:
```

Comment:
Wondering what's the best approach here when the pattern port is null.

For `https://example.com`, whether you specified it from a string or a pattern, `url pattern's port` will be null, which means we will match any port. If we don't want to have pattern-like behaviors, then maybe when `url pattern's port` is null, we only match when `url's port` is also null?

To me it seems like "matching any port" should be opt in, but maybe that's not how URLPattern works?

Reply:
> why port would be more special than any other part of the URL here?

Mostly because http and https have default ports, http with no port usually implies port 80, https with no port implies 443?

Reply:
Part of my problem is that while providing no port in the `pattern` type sounds intentional, if you use a string `https://example.com`, it's really hard to tell if you want to match any port or if you expect it to use the default for the protocol. So maybe we need a `has port` flag similar to the other parts?